### ISSUE TITLE
HLSL: crash while flattening

### ISSUE THREAD

**Issue by xxxbxxx** [KhronosGroup/glslang#954]
This crashes glslangValidator:
(uncommenting "float b" no longer crashes)

```
struct S1
{
	float4 a;
//	float b;
};

struct Param 
{
        Texture1D<float>        t;
	S1 			s1;
};

float func(in const Param In)
{
	const S1 s = In.s1;

	return s.a.x;
}

float4 main(float4 pos : SV_Position) : COLOR0
{
    Param p;
    p.s1.a = pos;

    float f = func(p);
    return float4(f);
}
```

**Comment by johnkslang** [comment]
I looked at this, and found out this much: The initializer

```
const S1 s = In.s1;
```

- correctly figures out it's assigning a non-flattened `S1 <- S1`
- has correctly replaced `In.s1` with the flattened shadow from Param being a flattened type
- incorrectly has (or incorrectly uses) a unique-id in the result `In.s1`, which is the same unique-id of something whose type was `Param`, rather than `S1`

The last point means when it looks at 

````
    const bool isFlattenRight = wasFlattened(right); // about line 2429
```

it thinks `right` was flattened as `Param` rather than seeing that `right` is a non-flattened `S1`.

Either the unique-id of `In.s1` is incorrect, or this situtaion should not be consuming the unique-id.

**Comment by johnkslang** [comment]
Update: This isn't a bug so much as a design hole, where the AST currently can't represent a partially, but not fully, dereferenced flattened object.  In the above example, a partial dereference is eventually mistaken as the full original object, and S1 gets conflated with Param (which is why it works better if they both have two members, but that's far from the point).

Will be working on a real fix....

**Comment by SebastienLussier** [comment]
Hello! I've got a similar instance of a flattening issue:

```
struct S1
{
    SamplerState samplerState;
};

struct S2
{
    S1 resources;
};

float4 main(float4 vpos : VPOS) : COLOR0
{
	S1 s1;
	S2 s2;
    s2.resources = s1;
    return float4(0,0,0,0);
}
```

Let me know if you need more details. This issue is pretty critical to us.
Thanks!

**Comment by johnkslang** [comment]
I have a pretty complete handling of subsets of flattened aggregates now.

The first part was previously done in d1be7545.

The remainder is working locally and should go in tomorrow. It coincidentally also involves fixing the problem fixed by #1081, so will look for just this one to fix it all rather than deal with conflicts.
